src/Aspire.Hosting/ApplicationModel/ResourceNotificationService.cs

```diff
@@ -13,12 +14,71 @@ namespace Aspire.Hosting.ApplicationModel;
 {
+    private static readonly TimeSpan s_defaultWaitForResourceStateTimeout = TimeSpan.FromSeconds(60);
     // Resource state is keyed by the resource and the unique name of the resource. This could be the name of the resource, or a replica ID.
     private readonly ConcurrentDictionary<(IResource, string), ResourceNotificationState> _resourceNotificationStates = new();
     private readonly ILogger<ResourceNotificationService> _logger = logger ?? throw new ArgumentNullException(nameof(logger));
+    private readonly CancellationToken _applicationStopping;
 
     private Action<ResourceEvent>? OnResourceUpdated { get; set; }
 
+    /// <summary>
+    /// Creates a new instance of <see cref="ResourceNotificationService"/>.
+    /// </summary>
+    /// <param name="logger">The logger.</param>
+    /// <param name="hostApplicationLifetime">The host application lifetime.</param>
+    public ResourceNotificationService(ILogger<ResourceNotificationService> logger, IHostApplicationLifetime hostApplicationLifetime)
+        : this(logger)
+    {
+        ArgumentNullException.ThrowIfNull(hostApplicationLifetime);
+
+        _applicationStopping = hostApplicationLifetime.ApplicationStopping;
+    }
+
+    /// <summary>
+    /// Waits for a resource to reach the specified state. See <see cref="KnownResourceStates"/> for common states.
+    /// </summary>
+    /// <remarks>
+    /// This method returns a task that will complete when the resource reaches the specified state. If the resource is already in the target state, the method will return immediately.<br/>
+    /// If the resource never reaches the target state, the method will wait for the passed <see cref="CancellationToken"/> to be cancelled, or 60 seconds if <c>null</c>, and then throw <see cref="TaskCanceledException"/>.
+    /// </remarks>
+    /// <param name="resourceName">The name of the resouce.</param>
+    /// <param name="targetState">The state to wait for the resource to transition to. See <see cref="KnownResourceStates"/> for common states. Defaults to <see cref="KnownResourceStates.Running"/>.</param>
+    /// <param name="cancellationToken">The cancellation token to monitor to cancel the wait operation. If <c>null</c> defaults to waiting for 60 seconds.</param>
+    /// <returns>A <see cref="Task"/> representing the wait operation.</returns>
+    public async Task WaitForResourceAsync(string resourceName, string? targetState = null, CancellationToken? cancellationToken = null)
+    {
+        var stateToWaitFor = targetState ?? KnownResourceStates.Running;
+        CancellationTokenSource? timeoutCts = null;
+        CancellationToken token;
+        if (cancellationToken is { } ct)
+        {
+            token = ct;
+        }
+        else
+        {
+            timeoutCts = new(s_defaultWaitForResourceStateTimeout);
+            token = timeoutCts.Token;
+        }
+        var cts = CancellationTokenSource.CreateLinkedTokenSource(_applicationStopping, token);
```

Comment:
Should we update `WatchAsync` to do this too so that it cancels when the `_applicationStopping` token is signalled?

Reply:
@davidfowl thoughts on this? Currently, the `IAsyncEnumerable` returned from `WatchAsync` does not get cancelled when the application stopping token is triggered. Should we fix that?

Reply:
Leave it where it is for now. We can always move it.